core/types/calls.cc

```diff
@@ -4141,6 +4141,64 @@ class Kernel_proc : public IntrinsicMethod {
     }
 } Kernel_proc;
 
+class Kernel_raise : public IntrinsicMethod {
+public:
+    void apply(const GlobalState &gs, const DispatchArgs &args, DispatchResult &res) const override {
+        if (args.args.size() < 1) {
+            return;
+        }
+
+        // The isUntyped check is part performance optimization and part "correctness":
+        // - no need to run another dispatch if it's just going to be untyped, but also...
+        // - arguably none of this intrinsic should run if we don't know that arg0 isn't a Class
+        //
+        // Technically speaking, the Ruby VM actually implements this by way of calling
+        // arg0.exception. It just so happens that Exception.exception is defined to forward to
+        // `.new` (https://ruby-doc.org/core-2.7.2/Exception.html#method-c-exception) but anything
+        // that defines a method called `exception` could be called. We don't implement that here,
+        // because implicit conversions are hard in the presence of subtyping. Instead, we restrict
+        // the check to only subclasses of `Exception`, not arbitrary classes.
+        auto classOfException = make_type<ClassType>(Symbols::Exception().data(gs)->lookupSingletonClass(gs));
+        if (!args.args[0]->type.isUntyped() && !Types::isSubType(gs, args.args[0]->type, classOfException)) {
```

Comment:
WDYT about pulling `args.args[0]` out into a separate variable (`klass` or somesuch) and using it here and below?